clang/lib/Driver/Driver.cpp

```diff
@@ -6440,7 +6440,8 @@ const ToolChain &Driver::getToolChain(const ArgList &Args,
       TC = std::make_unique<toolchains::NVPTXToolChain>(*this, Target, Args);
       break;
     case llvm::Triple::AMDHSA:
-      TC = std::make_unique<toolchains::ROCMToolChain>(*this, Target, Args);
+      TC = std::make_unique<toolchains::ROCMToolChain>(*this, Target, Args,
+                                                       Target.isOSWindows());
```

Comment:
Why do we need to pass something implied by the triple when we already have the triple? Also, this is for directly targeting via `--target=amdgcn-amd-amdhsa`, which isn't Windows.

Reply:
The reason we need to pass this is primarily/solely because when we create the Offloading Device ToolChain (in "Driver::getOffloadingDeviceToolChain()", we pass the Device's triple to the toolchain constructor as the target triple; which as you mentioned will not be Windows.  Then the Device side toolchain, has it's own RocmInstallationDetector, which will add the linux paths.  And I need a way to tell the RocmInstallationDetector, on the device side ToolChain, that the Host TC is actually Windows.  

I tried a couple of different approaches to handle this case, but this is the least intrusive so far.  The RocmInstallationDetector sets the search paths in its constructor, and all of the ToolChains that have a Rocm Detector call its constructor in their respective initialize list in their constructors.  I need a way to tell that detector instance (for a device side toolchain) that it's Host ToolChain is windows and to not add the linux search paths.

Reply:
I need to wrap my head around this more. We create multiple toolchains for offloading, the host toolchain and then the additional ones for offloading. All of those toolchains inherit from `Generic_GCC` in this case which initializes the detector. For the host toolchain this will just be the normal host triple, but for the GPU it should be the GPU target. Since we have the triple in the installation detector, we should just be able to change behavior if the OS is windows? We have the triple argument for ROCm we just don't use it, what's stopping you from just saving that triple inside the class and using it when we do our detection?

Reply:
"_I need to wrap my head around this more_"
Yup, the class hierarchy we have created for our tool chains is ... frustrating, for this case.  And I've tried a few different variations just to be able to tell the Device's ToolChain's RocmInstallationDetector to hold off on adding these search paths.

"_what's stopping you from just saving that triple inside the class and using it when we do our detection_"

That was one of my attempts.  I first tried just adding it to the RocmInstallationDetector (RID) class.  But because we use the LazyDetector template, I couldn't change the RID constructor without changing the LazyDetector template, and I didn't want to mess around with any of the other Detectors.

So, then I tried just adding a Triple member to the ROCMToolChain class for the Host Triple, and initializing it in its constructor. Which works.  But it is essentially ends up being the same (or nearly the same) as just passing a flag, and doing the "work" in the ROCMToolChain constructor.   If I use a member Triple for Host, I would still need to check it in the ROCMToolChain ctor, and tell its RID that the Host is Windows, before having the RID generate the search paths.

Reply:
The `RID` already has the host triple available I thought https://github.com/llvm/llvm-project/blob/3734e4c0c4966f794b9588445448004fee769ab9/clang/lib/Driver/ToolChains/ROCm.h#L169. Is the issue the fact that we probably make one for the GPU toolchain? If it's an unrecognized triple it should probably just give up.

Reply:
Right, we call it "HostTriple", but in the case for a Device side ToolChain, the triple passed to the TC's ctor (which is the same triple passed to the RID ctor) is going to be the device triple, which is always (I think) amdgcn-amd-amdhsa, or at least definitely not Windows.  We can use the TC's triple for removing the linux search paths for Host side.  But the problem is that this doesn't work when we create the Device TC, and it's RID.

Reply:
Do we actually require the ROCm installation from the device's TC? I guess it'd be used when making the device interface? The `HIPAMDToolChain` has a reference to the host toolchain, so I'd argue that we should use that one's ROCm installation and never use the device one.

Reply:
@jhuber6 So I spent some time trying to rework this to do just what you suggested (have the HIPAMDToolChain, just use the HostTC's RocmInstallation), because it is a good idea  :-)  But ... I ran into design issues.  The member "HostTC" of HIPAMDToolChain, is a "const ToolChain &".  And the "ToolChain" class does not have a "RocmInstallation" member, it is a member of the child class "Generic_GCC". We could cast "HostTC" reference to a Generic_GCC object reference (just in the HIPAMDToolClass), but that solution doesn't feel very pretty either.

Reply:
We do stuff like this in a *ton* of places,
```
clang/lib/Driver/ToolChains/Cuda.cpp
387:      static_cast<const toolchains::NVPTXToolChain &>(getToolChain());
533:      static_cast<const toolchains::CudaToolChain &>(getToolChain());
584:      static_cast<const toolchains::NVPTXToolChain &>(getToolChain());
```

Reply:
ah ... ok.  :-)